### PR TITLE
Optical keyboard

### PR DESCRIPTION
**Pull request by girishji**
Optical keyboard with Ergo layout

## Description

This is an optical matrix keyboard corresponding to pcb here: https://github.com/girishji/optical-keyboard. It uses Gateron optical switches with a blackpill stm32f401 mcu. There is also a custom matrix to drive the Led's and PT's.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
Lint has flagged a few issues.  Please resolve.

**Comment by girishji** [comment]
Fixed the lint issue

**Comment by girishji** [comment]
Thanks, I fixed and committed (it says it cannot generate info.json, and I
found a bug in the naming of keyboard to match dir name)

File: keyboards/opticalkb/blackpill_f401.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/* Copyright 2020 Sergey Vlasov (sigprof)
```
Comment (zvecr) [inline keyboards/opticalkb/blackpill_f401.c:1]: File should be deleted

File: keyboards/opticalkb/rules.mk [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+DYNAMIC_MACRO_ENABLE = yes  # Dynamic macros
+
+LED_MATRIX_ENABLE = no
+
+# turn this off after debugging
+# CONSOLE_ENABLE = yes
+
+# https://github.com/qmk/qmk_firmware/blob/master/docs/custom_matrix.md
+CUSTOM_MATRIX = yes
+SRC += matrix.c blackpill_f401.c
```
Comment (zvecr) [inline keyboards/opticalkb/rules.mk:33]:
```suggestion
SRC += matrix.c
```

File: keyboards/opticalkb/mcuconf.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/* Copyright 2020 QMK
```
Comment (zvecr) [inline keyboards/opticalkb/mcuconf.h:1]: File should be removed

File: keyboards/opticalkb/halconf.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/* Copyright 2020 QMK
```
Comment (zvecr) [inline keyboards/opticalkb/halconf.h:1]: File should be removed

File: keyboards/opticalkb/chconf.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/* Copyright 2020 QMK
```
Comment (zvecr) [inline keyboards/opticalkb/chconf.h:1]: File should be removed

**Review by zvecr — CHANGES_REQUESTED** [review]
* should use custom matrix "lite"
  * currently not implementing debounce correctly
  * unnecessary duplication of code
* info.json render does not match keyboard layout

File: keyboards/opticalkb/blackpill_f401.c (hunk 1 above)
Comment (girishji) [reply keyboards/opticalkb/blackpill_f401.c:1]: done

File: keyboards/opticalkb/rules.mk (hunk 2 above)
Comment (girishji) [reply keyboards/opticalkb/rules.mk:33]: done

File: keyboards/opticalkb/mcuconf.h (hunk 3 above)
Comment (girishji) [reply keyboards/opticalkb/mcuconf.h:1]: done

File: keyboards/opticalkb/chconf.h (hunk 5 above)
Comment (girishji) [reply keyboards/opticalkb/chconf.h:1]: done

**Comment by girishji** [comment]
> * should use custom matrix "lite"
>   
>   * currently not implementing debounce correctly

I am using an optical matrix. There is no need for debounce. I get stable switching even at very high scan rates (delay of 30us between reading columns). Debounce code is not applicable here. No need to keep 2 matrix copies and deal with timers. This will slow down scan rate.

>   * unnecessary duplication of code
The duplication is unavoidable unless code is refactored in matrix_common

> * info.json render does not match keyboard layout
I need some help here. How is this checked? I see 5x15 keys in info.json and it matches matrix definition

**Comment by girishji** [comment]
Hello Joel, I made the changes you requested and commited. I have some
questions, please see the github page.

**Comment by zvecr** [comment]
> I am using an optical matrix. Optical switches do not need debounce. It takes about 15us for phototransistor to rise and IO pin will read accurate values after that

You should still implement it correctly, but set the DEBOUNCE config option to zero.

**Comment by zvecr** [comment]
```
Layout Layout:
┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
│K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0│
└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
│K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1│
└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
│K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2│
└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
│K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3│
└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
│K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4│
└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
```
does not represent what is shown in

image

File: keyboards/opticalkb/readme.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# optkb
+
+*A short description of the keyboard/project*
```
Comment (drashna) [inline keyboards/opticalkb/readme.md:3]: Might want to fill this out

File: keyboards/opticalkb/readme.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+Verify that it entered DFU (bootloader) mode through QMK Toolbox before
+flashing.
+
+## Wiring instructions
+
+There is only one way to solder Blackpill (STM32F401) to the [PCB](https://github.com/girishji/optical-keyboard).
+
+#### Notes
+
+- On Blackpill boards, **avoid** using the following pins, since they will cause either USB enumeration or the DFU bootloader to not work correctly:
```
Comment (drashna) [inline keyboards/opticalkb/readme.md:36]: This is linked in the docs:
https://docs.qmk.fm/#/platformdev_blackpill_f411

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/opticalkb/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    // debug_enable=true;
+    // debug_matrix=true;
+    // debug_keyboard=true;
+    // debug_mouse=true;
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    // If console is enabled, it will print the matrix position and status of each key pressed
+#ifdef CONSOLE_ENABLE
+    uprintf("KL: kc: 0x%04X, col: %u, row: %u, pressed: %b, time: %u, interrupt: %b, count: %u\n", keycode, record->event.key.col, record->event.key.row, record->event.pressed, record->event.time, record->tap.interrupted, record->tap.count);
```
[the remark this answers is not in the document]
Comment (girishji) [reply keyboards/opticalkb/keymaps/default/keymap.c]: agree, I made the changes, thank you!

File: keyboards/opticalkb/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    switch (keycode) {
+        case QMKBEST:
+            if (record->event.pressed) {
+                // when keycode QMKBEST is pressed
+                SEND_STRING("QMK is the best thing ever!");
+            } else {
+                // when keycode QMKBEST is released
+            }
+            return false; // no more processing necessary
```
[the remark this answers is not in the document]
Comment (girishji) [reply keyboards/opticalkb/keymaps/default/keymap.c]: Bummer, removed that stuff

**Comment by girishji** [comment]
> ```
> Layout Layout:
> ┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
> │K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0││K0│
> └──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
> ┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
> │K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1││K1│
> └──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
> ┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
> │K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2││K2│
> └──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
> ┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
> │K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3││K3│
> └──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
> ┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐┌──┐
> │K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4││K4│
> └──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘└──┘
> ```
> 
> does not represent what is shown in
> 
> 

Thanks, changed it. I am not sure how I can verify this graphically though, but changed the geometry values

File: keyboards/opticalkb/readme.md (hunk 7 above)
Comment (girishji) [reply keyboards/opticalkb/readme.md]: Thank you

File: keyboards/opticalkb/readme.md (hunk 6 above)
Comment (girishji) [reply keyboards/opticalkb/readme.md]: Thank you, will add

File: keyboards/opticalkb/rules.mk [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+DYNAMIC_MACRO_ENABLE = yes  # Dynamic macros
+
+LED_MATRIX_ENABLE = no
+
+# turn this off after debugging
+# CONSOLE_ENABLE = yes
+
+# https://github.com/qmk/qmk_firmware/blob/master/docs/custom_matrix.md
```
[the remark this answers is not in the document]
Comment (girishji) [reply keyboards/opticalkb/rules.mk]: changed

**Comment by girishji** [comment]
> > I am using an optical matrix. Optical switches do not need debounce. It takes about 15us for phototransistor to rise and IO pin will read accurate values after that
> 
> You should still implement it correctly, but set the `DEBOUNCE` config option to zero.
> 
> > The duplication is unavoidable unless code is refactored in matrix_common
> 
> Using `lite` removes the duplication of various things, which is why its been suggested.

Thanks, changed as requested. I am glad setting DEBOUNCE to 0 eliminates all debouce related processing

**Comment by girishji** [comment]
Thank you, I think I have incorporated all suggestions

File: keyboards/opticalkb/rules.mk [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+CUSTOM_MATRIX = yes
```
[the remark this answers is not in the document]
Comment (girishji) [reply keyboards/opticalkb/rules.mk]: Done

**Comment by girishji** [comment]
Please review and merge, thanks :)

File: keyboards/opticalkb/rev1/config.h [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID 0xFEFD
+#define PRODUCT_ID 0x5302
+#define DEVICE_VER 0x0002
+#define MANUFACTURER girish
+#define PRODUCT opticalkb
```
Comment (drashna) [inline keyboards/opticalkb/rev1/config.h:13]: This needs to be migrated to the info.json file.

Comment (girishji) [reply keyboards/opticalkb/rev1/config.h]: Please review, I committed the requested changes.

File: keyboards/opticalkb/rev1/config.h [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+/* PT requires time to 'rise' after IR has been powered on. Read the GPIO input
+ * pin only after PT voltage has stabilized. */
+#define WAIT_AFTER_COL_SELECT 76
+/* PT requires time to 'fall' after IR has been switched off. Wait until
+ * voltage across PT has stabilized. Otherwise scanning of next column will be
+ * affected. */
+#define WAIT_AFTER_COL_UNSELECT 110
+
+// Enable 1000hz scan rate
+#define USB_POLLING_INTERVAL_MS 1
```
Comment (drashna) [inline keyboards/opticalkb/rev1/config.h:36]: Is enabled by default.
```suggestion
```

File: keyboards/opticalkb/rev1/config.h [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+// #define RGBLIGHT_LAYERS_OVERRIDE_RGB_OFF
+// #ifdef RGB_DI_PIN
+// #    define RGBLED_NUM 16
+// #    define RGBLIGHT_HUE_STEP 8
+// #    define RGBLIGHT_SAT_STEP 8
+// #    define RGBLIGHT_VAL_STEP 8
+// #    define RGBLIGHT_LIMIT_VAL 255 /* The maximum brightness level */
+// #    define RGBLIGHT_SLEEP  /* If defined, the RGB lighting will be switched off when the host goes to sleep */
+/*== all animations enable ==*/
+// #    define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/opticalkb/rev1/config.h:83]:
```suggestion
```

File: keyboards/opticalkb/rev1/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2022 Girish Palya (@girishji)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
+bool     is_cmd_grv_active = false;
```
Comment (drashna) [inline keyboards/opticalkb/rev1/keymaps/default/keymap.c:6]: Default keymap should be kept simple, as this is what is copied when using the new keymap command, and overly complex keymaps can be confusing for newer users. 

Creating a "stock" type keymap that has the more advanced features is fine. though.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/opticalkb/rev1/keymaps/default/keymap.c (hunk 15 above)
Comment (girishji) [reply keyboards/opticalkb/rev1/keymaps/default/keymap.c]: Thanks for suggestions. Requested changes are committed.

File: keyboards/opticalkb/rev1/rules.mk [hunk 16]
```diff
@@ -0,0 +1,30 @@
+# https://docs.qmk.fm/#/squeezing_avr?id=rulesmk-settings
+#
+
```
Comment (drashna) [inline keyboards/opticalkb/rev1/rules.mk:3]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/opticalkb/rev1/rules.mk (hunk 16 above)
Comment (girishji) [reply keyboards/opticalkb/rev1/rules.mk]: Thanks for suggestions. Requested changes are committed.

**Review by drashna — APPROVED** [review]

File: keyboards/opticalkb/rev1/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        OSM(MOD_LCTL), KC_ESC, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT, KC_RIGHT,
+        OSM(MOD_LSFT), KC_GRV, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, OSM(MOD_RSFT), KC_UP, KC_CAPS,
+        OSM(MOD_LALT), KC_BTN1, KC_BTN3, OSL(_FN), OSM(MOD_LGUI), KC_BSPC, OSM(MOD_RALT), KC_SPC, KC_PGDN, KC_PGUP, KC_HOME, KC_END, KC_LEFT, KC_DOWN, KC_RIGHT
+    ),
+    [_FN]   = LAYOUT(
+        _______, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9, KC_F10, KC_F11, KC_F12, KC_HOME, KC_END,
+        _______, DYN_REC_STOP, _______, _______, _______, _______, _______, _______, _______, _______, DYN_MACRO_PLAY1, _______, _______, _______, 
+        _______, _______, _______, DYN_REC_START1, _______, _______, _______, LCMD(KC_LEFT), KC_PGDN, KC_PGUP, LCMD(KC_RIGHT), _______, _______, _______, _______, 
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_WH_U, _______, 
+        RESET, KC_BTN2, _______, _______, _______, _______, _______, _______, KC_END, KC_HOME, _______, DYN_MACRO_PLAY1, _______, KC_WH_D, _______
```
Comment (fauxpark) [inline keyboards/opticalkb/rev1/keymaps/default/keymap.c:23]:
```suggestion
        QK_BOOT, KC_BTN2, _______, _______, _______, _______, _______, _______, KC_END, KC_HOME, _______, DYN_MACRO_PLAY1, _______, KC_WH_D, _______
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Pull request closed without merging**